evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
 	// Most evidence should be committed in the very next block that is why we wait
 	// just over the block production rate before sending evidence again.
 	broadcastEvidenceIntervalS = 10
-	// If a message fails wait this much before sending it again
-	peerRetryMessageIntervalMS = 100
 )
 
 // Reactor handles evpool evidence broadcasting amongst peers.
 type Reactor struct {
-	p2p.BaseReactor
-	evpool   *Pool
-	eventBus *types.EventBus
+	service.BaseService
+
+	evpool       *Pool
+	eventBus     *types.EventBus
+	evidenceCh   *p2p.Channel
+	peerUpdates  *p2p.PeerUpdatesCh
+	closeCh      chan struct{}
+	peerRoutines map[string]chan struct{}
+
+	mtx    tmsync.RWMutex
```

Comment:
What does this mutex protect? It is idiomatic to place the mtx directly above the fields that are protected by it.

Reply:
This is news to me. Can you show me some references that state that? It protects the `peerRoutines`.

Reply:
https://golang.org/doc/effective_go.html#commentary

> Grouping can also indicate relationships between items, such as the fact that a set of variables is protected by a mutex.

```go
var (
    countLock   sync.Mutex
    inputCount  uint32
    outputCount uint32
    errorCount  uint32
)
```

Reply:
One convention that I recently discovered, and quite like, is to use an inline struct for mutex-protected fields -- this is used e.g. in CockroachDB:

```go
type Foo struct {
  logger log.Logger

  mu struct {
    sync.Mutex
    shared string
  }
}

func (f *Foo) Set(s string) {
  f.mu.Lock()
  defer f.mu.Unlock()
  f.mu.shared = s
}
```

Reply:
Thanks!